README
== Money class

This money class is based on the example from the ActiveRecord doc:
http://api.rubyonrails.org/classes/ActiveRecord/Aggregations/ClassMethods.html

Its in production use at http://www.snowdevil.ca and I haven't found any major issues
so far. 
The main reason to open source it is because It might be useful to other people and 
I hope i'll get some feedback on how to improve the class. 

I bundled the exporter with the money class since some tests depend on it and I figured
that most applications which need to deal with Money also need to deal with proper 
exporting. 

== Download

Preferred method of installation is gem: 

  gem install --source http://dist.leetsoft.com money 

Alternatively you can get the library packed
  
  http://dist.leetsoft.com/pkg/

== Useage

Use the compose_of helper to let active record deal with embedding the money
object in your models. The following example requires a cents and a currency field.

  class ProductUnit < ActiveRecord::Base
    belongs_to :product
    composed_of :price, :class_name => "Money", :mapping => [%w(cents cents) %(currency currency)]

    private        
      validate :cents_not_zero
    
      def cents_not_zero
        errors.add("cents", "cannot be zero or less") unless cents > 0
      end
    
      validates_presence_of :sku, :currency
      validates_uniqueness_of :sku        
  end
  
== Class configuration

Two const class variables are available to tailor Money to your needs. 
If you don't need currency exchange at all, just ignore those.

=== Default Currency

By default Money defaults to USD as its currency. This can be overwritten using

  Money.default_currency = "CAD"
  
If you use rails, the environment.rb is a very good place to put this. 

=== Currency Exchange

The second parameter is a bit more complex. It lets you provide your own implementation of the 
currency exchange service. By default Money throws an exception when trying to call .exchange_to. 

A second minimalist implementation is provided which lets you supply custom exchange rates:

  Money.bank = VariableExchangeBank.new 
  Money.bank.add_rate("USD", "CAD", 1.24515)
  Money.bank.add_rate("CAD", "USD", 0.803115)
  Money.us_dollar(100).exchange_to("CAD") => Money.ca_dollar(124)
  Money.ca_dollar(100).exchange_to("USD") => Money.us_dollar(80)

There is nothing stopping you from creating bank objects which scrape www.xe.com for the current rates or just return rand(2)
  
== Code

If you have any improvements please email them to tobi [at] leetsoft.com 


CHANGELOG
CHANGELOG
---

* Removed nil.to_money. This would yield dangerous bugs..

* Added money.zero? to mimik Numeric.zero?

tests/money_test.rb
$:.unshift(File.dirname(__FILE__) + '/../lib')

require 'test/unit'
require 'money'

class MoneyTest < Test::Unit::TestCase
  
  def setup
    
    @can1  = Money.ca_dollar(100)
    @can2  = Money.ca_dollar(200)
    @can3  = Money.ca_dollar(300)
    @us1   = Money.us_dollar(100)
    
    Money.bank = NoExchangeBank.new
    Money.default_currency = "USD"
    
  end
  
  def test_sanity

    assert_equal @can1, @can1
    assert_equal true, @can1 < @can2
    assert_equal true, @can3 > @can2
    assert_equal [@can1, @can2, @can3], [@can1, @can2, @can3]
    assert_equal [@can1, @can2, @can3], [@can3, @can2, @can1].sort
    assert_not_equal [@can1, @can2, @can3], [@can3, @can2, @can1]
    assert_equal [@can1], [@can1, @can2, @can3] & [@can1]

    assert_equal [@can1, @can2], [@can1] | [@can2]

    
    assert_equal @can3, @can1 + @can2
    assert_equal @can1, @can2 - @can1
    
    assert_equal Money.ca_dollar(500), Money.ca_dollar(10) + Money.ca_dollar(90) + Money.ca_dollar(500) - Money.ca_dollar(100)
      
  end
  
  def test_default_currency
    
    assert_equal Money.new(100).currency, "USD"
    Money.default_currency = "CAD"
    assert_equal Money.new(100).currency, "CAD"
  end
  
  def test_default_exchange   
    assert_raise(Money::MoneyError) do
      Money.us_dollar(100).exchange_to("CAD")
    end   
  end
  
  def test_zero
    assert_equal true, Money.empty.zero?
    assert_equal false, Money.ca_dollar(1).zero?
    assert_equal false, Money.ca_dollar(-1).zero?
  end
  
  def test_real_exchange   
    Money.bank = VariableExchangeBank.new
    Money.bank.add_rate("USD", "CAD", 1.24515)
    Money.bank.add_rate("CAD", "USD", 0.803115)
    assert_equal Money.us_dollar(100).exchange_to("CAD"), Money.ca_dollar(124)
    assert_equal Money.ca_dollar(100).exchange_to("USD"), Money.us_dollar(80)
  end
    
  def test_multiply    
    assert_equal Money.ca_dollar(5500), Money.ca_dollar(100) * 55    
    assert_equal Money.ca_dollar(150), Money.ca_dollar(100) * 1.50
    assert_equal Money.ca_dollar(50), Money.ca_dollar(100) * 0.50
  end

  def test_divide
    assert_equal Money.ca_dollar(100), Money.ca_dollar(5500) / 55    
    assert_equal Money.ca_dollar(100), Money.ca_dollar(200) / 2    
  end
  
  def test_empty_can_exchange_currency
    assert_equal Money.ca_dollar(100), Money.empty('USD') + Money.ca_dollar(100)
    assert_equal Money.ca_dollar(100), Money.ca_dollar(100) + Money.empty('USD')
    
    assert_equal Money.ca_dollar(-100), Money.empty('USD') - Money.ca_dollar(-100)
    assert_equal Money.ca_dollar(-100), Money.ca_dollar(-100) - Money.empty('USD')
  end
  
  def test_to_s
    assert_equal "1.00", @can1.to_s
    assert_equal "390.50", Money.ca_dollar(39050).to_s
  end
  
  def test_formatting

    assert_equal "free", Money.ca_dollar(0).format
    assert_equal "$1.00", @can1.format 
    assert_equal "$1.00 CAD", @can1.format(:with_currency)

    assert_equal "$1", @can1.format(:no_cents)
    assert_equal "$5", Money.ca_dollar(570).format(:no_cents)

    assert_equal "$5 CAD", Money.ca_dollar(570).format([:no_cents, :with_currency])
    assert_equal "$5 CAD", Money.ca_dollar(570).format(:no_cents, :with_currency)
    assert_equal "$390", Money.ca_dollar(39000).format(:no_cents)

    assert_equal "$5.70 <span class=\"currency\">CAD</span>", Money.ca_dollar(570).format([:html, :with_currency])
    
  end
  
  
end

tests/core_extensions_test.rb
$:.unshift(File.dirname(__FILE__) + '/../lib')

require 'test/unit'
require 'money'

class CoreExtensionsTest < Test::Unit::TestCase

  def setup
  end

  def test_numeric_conversion    
    assert_equal Money.new(10000), 100.to_money
    assert_equal Money.new(10038), 100.38.to_money
    assert_equal Money.new(-10000), -100.to_money
  end
    
  def test_string_conversion
    assert_equal Money.new(100), "$1".to_money
    assert_equal Money.new(100), "$1.00".to_money
    assert_equal Money.new(137), "$1.37".to_money
    assert_equal Money.new(100, 'CAD'), "CAD $1.00".to_money
    assert_equal Money.new(-10000), "-100".to_money
    assert_equal Money.new(410), "4.10".to_money
  end
  
  def test_nil
    assert_raise(NoMethodError) do
      nil.to_money
    end
  end

end

lib/money/core_extensions.rb
# Allows Writing of 100.to_money for +Numeric+ types
#   100.to_money => #<Money @cents=10000>
#   100.37.to_money => #<Money @cents=10037>
class Numeric
  def to_money
    Money.new(self * 100)
  end
end

# Allows Writing of '100'.to_money for +String+ types
# Excess characters will be discarded
#   '100'.to_money => #<Money @cents=10000>
#   '100.37'.to_money => #<Money @cents=10037>
class String
  def to_money
    # Get the currency
    matches = scan /([A-Z]{2,3})/ 
    currency = matches[0] ? matches[0][0] : Money.default_currency
    
    # Get the cents amount
    matches = scan /(\-?\d+(\.(\d+))?)/
    cents = matches[0] ? (matches[0][0].to_f * 100).round : 0
    
    Money.new(cents, currency)
  end
end

# Allow writing of Nil.to_money => Money.empty
class NilClass
  def to_money
    Money.empty
  end  
end

lib/money/money.rb
# === Usage with ActiveRecord
# 
# Use the compose_of helper to let active record deal with embedding the money
# object in your models. The following example requires a cents and a currency field.
# 
#   class ProductUnit < ActiveRecord::Base
#     belongs_to :product
#     composed_of :price, :class_name => "Money", :mapping => [ %w(cents cents), %w(currency currency) ]
# 
#     private        
#       validate :cents_not_zero
#     
#       def cents_not_zero
#         errors.add("cents", "cannot be zero or less") unless cents > 0
#       end
#     
#       validates_presence_of :sku, :currency
#       validates_uniqueness_of :sku        
#   end
#   
class Money
  include Comparable

  attr_reader :cents, :currency

  class MoneyError < StandardError# :nodoc:
  end

  # Bank lets you exchange the object which is responsible for currency
  # exchange. 
  # The default implementation just throws an exception. However money
  # ships with a variable exchange bank implementation which supports
  # custom excahnge rates:
  #
  #  Money.bank = VariableExchangeBank.new
  #  Money.bank.add_rate("USD", "CAD", 1.24515)
  #  Money.bank.add_rate("CAD", "USD", 0.803115)
  #  Money.us_dollar(100).exchange_to("CAD") => Money.ca_dollar(124)
  #  Money.ca_dollar(100).exchange_to("USD") => Money.us_dollar(80)
  @@bank = NoExchangeBank.new
  cattr_accessor :bank

  @@default_currency = "USD"
  cattr_accessor :default_currency

  # Creates a new money object. 
  #  Money.new(100) 
  # 
  # Alternativly you can use the convinience methods like 
  # Money.ca_dollar and Money.us_dollar 
  def initialize(cents, currency = default_currency)
    @cents, @currency = cents.round, currency
  end

  # Do two money objects equal? Only works if both objects are of the same currency
  def eql?(other_money)
    cents == other_money.cents && currency == other_money.currency
  end

  def <=>(other_money)
    if currency == other_money.currency
      cents <=> other_money.cents
    else
      cents <=> other_money.exchange_to(currency).cents
    end
  end

  def +(other_money)
    return other_money.dup if cents.zero? 
    return dup if other_money.cents.zero?

    if currency == other_money.currency
      Money.new(cents + other_money.cents, other_money.currency)
    else
      Money.new(cents + other_money.exchange_to(currency).cents,currency)
    end   
  end

  def -(other_money)
    return other_money.dup if cents.zero? 
    return dup if other_money.cents.zero?

    if self.cents == 0 or currency == other_money.currency
      Money.new(cents - other_money.cents, other_money.currency)
    else
      Money.new(cents - other_money.exchange_to(currency).cents, currency)
    end   
  end

  # get the cents value of the object
  def cents
    @cents.to_i
  end

  # multiply money by fixnum
  def *(fixnum)
    Money.new(cents * fixnum, currency)    
  end

  # divide money by fixnum
  def /(fixnum)
    Money.new(cents / fixnum, currency)    
  end
  
  # Test if the money amount is zero
  def zero?
    cents == 0 
  end


  # Format the price according to several rules
  # Currently supported are :with_currency, :no_cents and :html
  #
  # with_currency: 
  #
  #  Money.ca_dollar(0).format => "free"
  #  Money.ca_dollar(100).format => "$1.00"
  #  Money.ca_dollar(100).format(:with_currency) => "$1.00 CAD"
  #  Money.us_dollar(85).format(:with_currency) => "$0.85 USD"
  #
  # no_cents:  
  #
  #  Money.ca_dollar(100).format(:no_cents) => "$1"
  #  Money.ca_dollar(599).format(:no_cents) => "$5"
  #  
  #  Money.ca_dollar(570).format(:no_cents, :with_currency) => "$5 CAD"
  #  Money.ca_dollar(39000).format(:no_cents) => "$390"
  #
  # html:
  #
  #  Money.ca_dollar(570).format(:html, :with_currency) =>  "$5.70 <span class=\"currency\">CAD</span>"
  def format(*rules)
    return "free" if cents == 0

    rules = rules.flatten

    if rules.include?(:no_cents)
      formatted = sprintf("$%d", cents.to_f / 100  )          
    else
      formatted = sprintf("$%.2f", cents.to_f / 100  )      
    end

    if rules.include?(:with_currency)
      formatted << " "
      formatted << '<span class="currency">' if rules.include?(:html)
      formatted << currency
      formatted << '</span>' if rules.include?(:html)
    end
    formatted
  end

  # Money.ca_dollar(100).to_s => "1.00"
  def to_s
    sprintf("%.2f", cents.to_f / 100  )
  end

  # Recieve the amount of this money object in another currency   
  def exchange_to(other_currency)
    self.class.bank.reduce(self, other_currency)
  end  

  # Create a new money object with value 0
  def self.empty(currency = default_currency)
    Money.new(0, currency)
  end

  # Create a new money object using the Canadian dollar currency
  def self.ca_dollar(num)
    Money.new(num, "CAD")
  end

  # Create a new money object using the American dollar currency
  def self.us_dollar(num)
    Money.new(num, "USD")
  end

  # Create a new money object using the Euro currency
  def self.euro(num)
    Money.new(num, "EUR")
  end

  # Recieve a money object with the same amount as the current Money object
  # in american dollar 
  def as_us_dollar
    exchange_to("USD")
  end

  # Recieve a money object with the same amount as the current Money object
  # in canadian dollar 
  def as_ca_dollar
    exchange_to("CAD")
  end

  # Recieve a money object with the same amount as the current Money object
  # in euro
  def as_ca_euro
    exchange_to("EUR")
  end  

  # Conversation to self
  def to_money
    self
  end  
end